src/test/java/com/google/sps/servlets/UserDataServletTest.java

```diff
@@ -0,0 +1,256 @@
+    writer.flush();
+
+    // Build the correct response output
+    Map<String, Object> correctResponse = new HashMap<>();
+    correctResponse.put(USER_FOUND_PROPERTY, false);
+
+    // Check that the user wasn't found in datastore
+    verify(mockRequest, atLeast(1)).getParameter(USER_ID_PROPERTY);
+    Assert.assertEquals(gson.toJson(correctResponse), stringWriter.toString().trim());
+  }
+
+  /**
+   * Tests the doGet method, where datastore does contain the test user id.
+   * Expected output: JSON response with "user-found" property as true,
+   * and the user properties filled in with the test user properties.
+   */
+  @Test
+  public void testGetMethodWithValidId() throws Exception {
+    // Mock the getParameter call
+    when(mockRequest.getParameter(USER_ID_PROPERTY)).thenReturn(TEST_USER_ID);
+
+    // Store output given by mockResponse
+    stringWriter = new StringWriter();
+    writer = new PrintWriter(stringWriter);
+    when(mockResponse.getWriter()).thenReturn(writer);
+    when(mockResponse.getContentType()).thenReturn("application/json");
+
+    // Create a local datastore
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    addTestUserEntityToDatastore(datastore);
+    // Test the doGet method by calling the wrapper method
+    servletUnderTest.doGetWrapper(datastore, mockRequest, mockResponse);
+    writer.flush();
+
+    // Build the correct response output
+    Map<String, Object> correctResponse = new HashMap<>();
+    correctResponse.put(USER_FOUND_PROPERTY, true);
+    correctResponse.put(USER_NAME_PROPERTY, TEST_USER_NAME);
+    correctResponse.put(USER_ID_PROPERTY, TEST_USER_ID);
+    correctResponse.put(USER_EMAIL_PROPERTY, TEST_USER_EMAIL);
+    correctResponse.put(USER_BIO_PROPERTY, TEST_USER_BIO);
+    correctResponse.put(USER_FRIENDS_LIST_PROPERTY, Arrays.asList(TEST_USER_FRIENDS_LIST));
+
+    // Check that the user was found in Datastore
+    verify(mockRequest, atLeast(1)).getParameter(USER_ID_PROPERTY);
+    Assert.assertEquals(gson.toJson(correctResponse), stringWriter.toString().trim());
+  }
+
+  /**
+   * Tests the doPost method, making sure it redirects correctly.
+   * Expected response: Redirects to the profile page with the user id.
+   */
+
+  @Test
+  public void testPostMethodRedirect() throws Exception {
+    // Mock the getParameter and getParameterValues calls
+    when(mockRequest.getParameter(USER_ID_PROPERTY)).thenReturn(TEST_USER_ID);
+    when(mockRequest.getParameter(USER_NAME_PROPERTY)).thenReturn(TEST_USER_NAME);
+    when(mockRequest.getParameter(USER_EMAIL_PROPERTY)).thenReturn(TEST_USER_EMAIL);
+    when(mockRequest.getParameter(USER_BIO_PROPERTY)).thenReturn(TEST_USER_BIO);
+    when(mockRequest.getParameterValues(USER_FRIENDS_LIST_PROPERTY)).thenReturn(TEST_USER_FRIENDS_LIST);
+
+    // Create a local datastore
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
```

Comment:
Random thought: Since you are setting the `servletUnderTest` in the `setUp`, should you maybe do the datastore there too?

Reply:
That makes sense. I made the change. Unfortunately there doesn't seem to be any way to close the datastore upon tear down, however.